rgw/notification: Fix the caching issues of notification brokers, where the cache was not invalidated if topic attributes were changed

kchheda3 opened the pull request:
Fixes https://tracker.ceph.com/issues/66036

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

yuvalif commented on src/rgw/driver/rados/rgw_notify.cc:
```diff
@@ -240,8 +240,11 @@ class Manager : public DoutPrefixProvider {
+  EntryProcessingResult process_entry(const ConfigProxy &conf,
+                                      persistency_tracker &entry_persistency_tracker,
+                                      const cls_queue_entry &entry,
+                                      RGWPubSubEndpoint *push_endpoint,
```
nit: `const RGWPubSubEndpoint* push_endpoint`

kchheda3 replied:
done.

yuvalif commented on src/rgw/driver/rados/rgw_notify.cc:
```diff
@@ -370,6 +363,51 @@ class Manager : public DoutPrefixProvider {
+  int create_push_endpoint(const std::string &queue_name,
+                           const std::vector<cls_queue_entry> &entries,
+                           spawn::yield_context yield,
+                           RGWPubSubEndpoint::Ptr *push_endpoint) {
```
why passing a pointer to a unique_ptr?
you can just pass the bare pointer from the unique_ptr

yuvalif replied:
or even better. since the caller ignore the return value from this function, you can return the unique_ptr.
you can return a unique_ptr initialized with nullptr in case of error

kchheda3 replied:
the caller does honor the return value, we do not process the entries if we cannot create the push_endpoint.
and i am passing the unique_ptr by ref and its created inside the function.
i do not want to verify based on whether the function returned nullptr

yuvalif replied:
ok. so please pass by ref, not a pointer. IMO, pointer to a unique_ptr is too confusing.

kchheda3 replied:
remove the code to pass unique_ptr

yuvalif commented on src/rgw/driver/rados/rgw_notify.cc:
```diff
@@ -446,7 +484,11 @@ class Manager : public DoutPrefixProvider {
+      RGWPubSubEndpoint::Ptr push_endpoint;
+      if (create_push_endpoint(queue_name, entries, yield, &push_endpoint)
+          < 0) {
+        return;
```
this means that you stop processing of the queue. i think we should continue queue processing unless we lose ownership of queue is removed.
this will alllow users to fix conf errors

yuvalif replied:
if you exit from the function, nothing will reschedule the processing of the queue and the rgw will release ownership.
handling should be similar to the other error cases (="continue"), and wait for the problem to be resolved.
note that we will not be scheduling the coroutine too often in this case, with the timer we have in line 426.

kchheda3 replied:
done

yuvalif commented on src/rgw/driver/rados/rgw_notify.cc:
```diff
@@ -461,10 +503,14 @@ class Manager : public DoutPrefixProvider {
-            auto result = process_entry(this->get_cct()->_conf, persistency_tracker, entry, yield);
+          auto result =
+              process_entry(this->get_cct()->_conf, persistency_tracker,
+                            entry, push_endpoint, yield);
```
how does this work?
`process_entry()` expect a bare pointer, not a unique_ptr. shouldn't that be:
```
 process_entry(this->get_cct()->_conf, persistency_tracker,
                            entry, push_endpoint.get(), yield);
```

yuvalif commented on src/rgw/driver/rados/rgw_notify.cc:
```diff
@@ -283,30 +286,20 @@ class Manager : public DoutPrefixProvider {
                         << " current time: " << time_now << dendl;
-    try {
-      // TODO move endpoint creation to queue level
-      const auto push_endpoint = RGWPubSubEndpoint::create(event_entry.push_endpoint, event_entry.arn_topic,
```
eventually we shoud remove these fields from the event entry.
but for mixed cluster compatibility, we should just keep them

kchheda3 replied:
> eventually we shoud remove these fields from the event entry. but for mixed cluster compatibility, we should just keep them

exactly, thats what we discussed, we keep them for now for compatibility and also populate them for now and we use them if for some reason we are not able to load the topic

yuvalif commented on src/rgw/rgw_amqp.cc:
```diff
@@ -53,14 +53,29 @@ static const int RGW_AMQP_NO_REPLY_CODE =                 0x0;
 // so, strings are copied into connection_id_t
-connection_id_t::connection_id_t(const amqp_connection_info& info, const std::string& _exchange)
-    : host(info.host), port(info.port), vhost(info.vhost), exchange(_exchange), ssl(info.ssl) {}
+connection_id_t::connection_id_t(
```
why changing the amqp code?

kchheda3 replied:
coz amqp too faces same issue, if for some reason there is typo in the username and password, coz the connection is pooled, the new connection never uses the new userid and passwd

yuvalif replied:
amqp code needs a major rewrite (there are several issues fixed for kafka that still exist in amqp).
also, we would need to test that.
would recommend splitting amqp work to another PR

kchheda3 replied:
i am afraid this will be forgotten while we refactor the amqp
since we are doing this for kafka, why not just do it for amqp as well, coz amqp also faces the same issue ? and its similar code changes so its logical to be part of same commit ?

yuvalif replied:
agree that this is needed there. but there are many other (critical) fixes we did on the kafka code that are needed there as well.
main reason that i want it in a separate PR is testing.

kchheda3 replied:
done, removed amqp code.

yuvalif commented on src/rgw/rgw_kafka.h:
```diff
@@ -21,28 +21,47 @@ bool init(CephContext* cct);
 void shutdown();
 
+// key class for the connection list
+struct connection_id_t {
```
can you move this to the cc file and just keep forward declaration here?

kchheda3 replied:
this struct is required in the `rgw_pubsub_push.cc` and i have followed the same logic as we have in `rgw_amqp.h`
any reason to move this to `.cc`?

yuvalif commented:
please add tests for both fixes:
* for the notification caching fix, you can define an invalid kafka port or host on persistent notifications, and then modify the topic to the correct values
* fot the connection caching, you can define 2 topics pointing to the same kafka broker, one with the `with-ssl` and one without. ans define the `with-ssl` topic first.
before your fix, all of the notifications to both topics would fail (as they share the same connection that tries to open a secure connection with the broker). after the fix, only notifications fro mthe first topics would fail, and the second topic should be successfull

yuvalif commented on src/rgw/driver/rados/rgw_pubsub_push.cc:
```diff
@@ -342,7 +338,7 @@ class RGWPubSubKafkaEndpoint : public RGWPubSubEndpoint {
   std::string to_str() const override {
     std::string str("Kafka Endpoint");
-    str += "\nBroker: " + conn_name;
+    str += "\nBroker: " + conn_id.broker;
```
```suggestion
    str += "\nBroker: " + to_string(conn_id);
```

yuvalif commented:
please rebase and resolve conflicts only after: https://github.com/ceph/ceph/pull/57630
is merged (as it will prevent you from testing kafka code).

kchheda3 commented:
> please rebase and resolve conflicts only after: https://github.com/ceph/ceph/pull/57630
> 
> is merged (as it will prevent you from testing kafka code).

Yes I was planning to take one final look at the #57630, and also do some local test again

kchheda3 commented:
> please add tests for both fixes:
> 
> * for the notification caching fix, you can define an invalid kafka port or host on persistent notifications, and then modify the topic to the correct values
> * fot the connection caching, you can define 2 topics pointing to the same kafka broker, one with the `with-ssl` and one without. ans define the `with-ssl` topic first.
>   before your fix, all of the notifications to both topics would fail (as they share the same connection that tries to open a secure connection with the broker). after the fix, only notifications fro mthe first topics would fail, and the second topic should be successfull

added both test cases.

kchheda3 commented:
> please rebase and resolve conflicts only after: #57630 is merged (as it will prevent you from testing kafka code).

rebased and updated the commits.

yuvalif commented:
since there is a behavior change here, please update: https://github.com/ceph/ceph/blob/main/PendingReleaseNotes
saying that after the fix:
* change to topics that were modified while persistent notifications were in the queue will be reflected in notifications
* different kafka topics that point to the same broker but that have different security parameters, will not override one another. instead, multiple connections with the appropriate security parameters will be used

yuvalif commented:
another small comment, IMO, you should try and avoid reformattign the code as you make changes to it.
regardless of which formatting is better... it make the review process more difficult.
if you really have the current formatting :-) please separate to a different commit in the PR.

kchheda3 commented:
> another small comment, IMO, you should try and avoid reformattign the code as you make changes to it. regardless of which formatting is better... it make the review process more difficult. if you really have the current formatting :-) please separate to a different commit in the PR.

yeah i only try to modify the code that i am touching
i usually use the editor for development which is configured to auto format the code as per google style as recommended in the [ceph](https://github.com/ceph/ceph/blob/main/CodingStyle#L21) 
sorry for the trouble

kchheda3 commented:
> since there is a behavior change here, please update: https://github.com/ceph/ceph/blob/main/PendingReleaseNotes saying that after the fix:
> 
> * change to topics that were modified while persistent notifications were in the queue will be reflected in notifications
> * different kafka topics that point to the same broker but that have different security parameters, will not override one another. instead, multiple connections with the appropriate security parameters will be used

was looking at the comments, do not see anything there specific to version v2, so wondering when do we change this release notes.
also since this is a bug fix, do we need to add generic release notes ?

kchheda3 commented:
> > > since there is a behavior change here, please update: https://github.com/ceph/ceph/blob/main/PendingReleaseNotes saying that after the fix:
> > > 
> > > * change to topics that were modified while persistent notifications were in the queue will be reflected in notifications
> > > * different kafka topics that point to the same broker but that have different security parameters, will not override one another. instead, multiple connections with the appropriate security parameters will be used
> > 
> > 
> > was looking at the comments, do not see anything there specific to version v2, so wondering when do we change this release notes.
> 
> you got me worried... but it is there! https://github.com/ceph/ceph/blob/main/PendingReleaseNotes#L26
> 
> > also since this is a bug fix, do we need to add generic release notes ?
> 
> usually we don't add bug fixes to the releas ewnotes, unless someone may expect the buggy behavior. not 100% sure about this fix. we probably don't need to put a note about the kafka connections (i don't see how anyone would want to override the values...), but i think that we probably do want to add a note about the persistent nmotifications. users maye had woraround practices for this case (where they used wrong parameters fo the broker), like deleting the topic and the queue, IMO, we should notify them that this is not needed anymore.

add release notes

yuvalif commented:
this PR has 2 fixes in it:
* in the kafka code, differentiating between 2 endpoints according to all of the endpoint parameters, not just hostname and port
* in the persistent notification code, taking the endpoint definitions from the topic config object instead of the infromation stored together with the notification entry in the queue.
an alternative solution for the same issue (=stale topic information in the queue that keeps retrying towards an invalid endpoint) could be found here: https://github.com/ceph/ceph/pull/5826
the main issue that the alternative approch is addressing, is that the topic configiration object name did not contain the tenant name prior to `squid`.
in the alternative approach, the endpoint data is stored in the "head" of the queue, this allows updates when topic endpoint is updated.

tried to compare the two approches:
* in both approaches, if the queue was created before `squid`, we will have to fallback to the existing approach of taking the data cached in the entry
* in the alternative approach we save on access to the topic object. however, this should be amortized over many calls to `process_entry()` so should not have significant perf cost
* in the alternative approach, we may add "migration" of a pre-dquid queue, by adding the relevant information to the head if it is missing
* the approach in this PR is a smaller change to the codebase - which is always better :-)

cbodley commented:
all of the notification jobs failed in https://pulpito.ceph.com/cbodley-2024-06-28_14:14:22-rgw-wip-cbodley-testing-distro-default-smithi/

i'm not sure whether the failures are due to this pr or https://github.com/ceph/ceph/pull/58081, which was in the same batch

@yuvalif can you please review the results?

yuvalif commented:
> all of the notification jobs failed in https://pulpito.ceph.com/cbodley-2024-06-28_14:14:22-rgw-wip-cbodley-testing-distro-default-smithi/
> 
> i'm not sure whether the failures are due to this pr or #58081, which was in the same batch
> 
> @yuvalif can you please review the results?

* `test_metadata_filter_ampq` failed - this is a new failure
* `test_lifecycle_http` and `test_lifecycle_kafka` failed - this is a new failure
* `test_ps_s3_persistent_topic_stats` failed - an expected failure

need to dig deeper to see what is the root cause

cbodley commented:
> need to dig deeper to see what is the root cause

thanks @yuvalif! in the meantime, i've pushed this PR and https://github.com/ceph/ceph/pull/58081 to separate ceph-ci branches so we can compare their rgw/notifications results

yuvalif commented:
probably unrelated, but amqp test has valgrind leak issues:
```
librados::v14_2_0::Rados::aio_create_completion()
rgw::notify::publish_commit()
rgw::sal::RadosNotification::publish_commit(()
RGWCopyObj/RGWCompleteMultipart/RGWPutObj::execute()
...
```
test seems to fail due to amqp broker issue. from rgw log:
```
new connection 'amqp://localhost:5672/?exchange=ex1' is created. but state creation failed (will retry). error: RGW_AMQP_STATUS_SOCKET_OPEN_FAILED (-9)
```
this means that `amqp_socket_open()` failed with -9. probably some environment or broker issue. IMO, should ere-run

yuvalif commented:
kafka failure also seems like an environment/broker issue. from rgw log:
```
RDKAFKA-3-FAIL: rdkafka#producer-1: [thrd:localhost:9092/bootstrap]: localhost:9092/bootstrap: Connect to ipv4#127.0.0.1:9092 failed: Connection refused (after 351ms in state CONNECT)
```

cbodley commented:
results from this pr alone: https://pulpito.ceph.com/cbodley-2024-07-01_19:02:58-rgw:notifications-wip-66036-distro-default-smithi/

yuvalif commented:
> results from this pr alone: https://pulpito.ceph.com/cbodley-2024-07-01_19:02:58-rgw:notifications-wip-66036-distro-default-smithi/

one failure in `test_ps_s3_persistent_topic_stats` is expected
